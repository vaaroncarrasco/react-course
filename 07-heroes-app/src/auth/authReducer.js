import { types } from '../types/types';

// * authReducer is just a pure function
export const authReducer = (state = {}, action) => {

    // const state = {
    //     name: 'Fernando',
    //     logged: true
    // }

    switch (action.type) {
        case types.login:
            return {
                ...action.payload,
                logged: true
            }

        case types.logout:
            return {
                logged: false
            }

        default:
            return state;
    }

}