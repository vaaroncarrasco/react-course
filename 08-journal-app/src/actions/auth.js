import Swal from "sweetalert2";

import { firebase, googleAuthProvider } from "../firebase/firebase-config";
import { types } from "../types/types"
import { noteLogout } from "./notes";
import { finishLoading, startLoading } from "./ui";

export const startLoginEmailPassword = (email, password) => {
    return (dispatch) => {

        dispatch( startLoading() );

        return firebase.auth().signInWithEmailAndPassword(email, password)
            .then( ( { user } ) => {

                dispatch(
                    login( user.uid, user.displayName )
                )

                dispatch( finishLoading() );


            }).catch( err => {
                dispatch( finishLoading() );
                console.log(err);
                Swal.fire('Error', err.message, 'error');
            });

    }
}


export const startRegisterWithEmailPasswordName = ( email, password, name ) => {

    // ? async fns return (callbacks) => {}
    return ( dispatch ) => {

        firebase.auth().createUserWithEmailAndPassword( email, password )
            .then( async ( {user} ) => {

                await user.updateProfile({ displayName: name });

                console.log(user)


                dispatch(
                    login( user.uid, user.displayName )
                )

            }).catch( err => {
                console.log(err);
                Swal.fire('Error', err.message, 'error');
            });

    }

}


export const startGoogleLogin = () => {

    return (dispatch) => {

        firebase.auth().signInWithPopup( googleAuthProvider )
            .then( ( { user } ) => {
                dispatch(
                    login( user.uid, user.displayName )
                )
            });

    }

}

export const login = (uid, displayName) => ({
    type: types.login,
    payload: {
        uid,
        displayName
    }
})

export const startLogout = () => {

    // ? Returns an async callback

    return async (dispatch) => {
        await firebase.auth().signOut();

        dispatch( logout() );

        dispatch( noteLogout() );
    }
}

export const logout = () => ({
    type: types.logout
})