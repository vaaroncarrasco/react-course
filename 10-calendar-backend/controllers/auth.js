const { response } = require('express');
const bcrypt = require('bcryptjs');
const Usuario = require('../models/Usuario');
const { generarJWT } = require('../helpers/jwt');

const crearUsuario = async( req, res = response ) => {

    const { email, password } = req.body;

    try {

        let usuario = await Usuario.findOne({ email });

        if ( usuario ) {
            return res.status(400).json({
                ok: false,
                msg: 'User already exists'
            });
        }

        usuario = new Usuario( req.body );

        // * encrypt password before saving to db
        const salt = bcrypt.genSaltSync();
        usuario.password = bcrypt.hashSync( password, salt );

        await usuario.save();

        // Generate JWT
        const token = await generarJWT( usuario.id, usuario.name );

        res.status(201).json({
            ok: true,
            msg: 'registro',
            uid: usuario.id,
            name: usuario.name,
            token
        });

    } catch (error) {
        console.log(error);
        res.status(500).json({
            ok: false,
            msg: 'hable con el admin'
        });
    }
}

const loginUsuario = async ( req, res = response ) => {

    const { email, password } = req.body;

    try {

        const usuario = await Usuario.findOne({ email });

        if ( !usuario ) {
            return res.status(400).json({
                ok: false,
                msg: 'No user found with email'
            });
        }

        // Confirm pass
        const validPassword = bcrypt.compareSync( password, usuario.password );

        if ( !validPassword ) {
            return res.status(400).json({
                ok: false,
                msg: 'Password incorrecto'
            });
        }

        // Generate JWT
        const token = await generarJWT( usuario.id, usuario.name );

        res.json({
            ok: true,
            uid: usuario.id,
            name: usuario.name,
            token
        })



    } catch (error) {
        console.log(error);
        res.status(500).json({
            ok: false,
            msg: 'hable con el admin'
        });
    }

}


const revalidarToken = async( req, res = response ) => {

    const { uid, name } = req;

    // * generate new JWT
    const token = await generarJWT( uid, name );

    res.json({
        ok: true,
        token,
        uid,
        name
    });
}

module.exports = {
    crearUsuario,
    loginUsuario,
    revalidarToken
}