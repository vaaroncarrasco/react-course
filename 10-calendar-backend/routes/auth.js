// * Auth / Users - Route | host + /api/auth

// node uses same lib it has in memory without importing it twice
const { Router } = require('express');
const { check } = require('express-validator');
const { validarCampos } = require('../middlewares/validar-campos');
const { validarJWT } = require('../middlewares/validar-jwt');
const router = Router();


const {
    crearUsuario,
    loginUsuario,
    revalidarToken
} = require('../controllers/auth');


router.post(
    '/new',
    [ // middlewares
        check('name', 'Name is required').not().isEmpty(),
        check('email', 'Email is required').isEmail(),
        check('password', 'Pass must be at least 6 character long').isLength({ min: 6 }),
        validarCampos
    ],
    crearUsuario
);

router.post(
    '/',
    [ // middlewares
        check('email', 'Email is required').isEmail(),
        check('password', 'Pass must be at least 6 character long').isLength({ min: 6 }),
        validarCampos
    ],
    loginUsuario
);

router.get('/renew', validarJWT, revalidarToken);

module.exports = router;