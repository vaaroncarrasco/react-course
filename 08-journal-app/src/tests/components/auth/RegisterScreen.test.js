import React from 'react';
import { mount } from "enzyme";
import { Provider } from "react-redux";
import { MemoryRouter } from 'react-router';
import { RegisterScreen } from '../../../components/auth/RegisterScreen';


import configureStore from 'redux-mock-store' //ES6 modules
import thunk from 'redux-thunk';
import '@testing-library/jest-dom';
import { types } from '../../../types/types';


const middlewares = [ thunk ];
const mockStore = configureStore(middlewares);


const initState = {
    auth: {},
    ui: {
      loading: false,
      msgError: null
    }
}

let store = mockStore(initState);
// we only simulate dispatch when we dont need it
// store.dispatch = jest.fn();

const wrapper = mount(
    <Provider store={ store }>
        <MemoryRouter>
            <RegisterScreen />
        </MemoryRouter>
    </Provider>
)

describe('<RegisterScreen />', () => {

    test('should match snapshot', () => {

        expect( wrapper ).toMatchSnapshot();

    });

    test('should dispatch respective action', () => {

        // ? .find( tag[attribute=value] )
        const emailField = wrapper.find('input[name="email"]');

        emailField.simulate('change', {
            target: {
                value: '',
                name: 'email'
            }
        });

        wrapper.find('form').simulate('submit', {
            preventDefault(){}
        });

        const actions = store.getActions();

        expect( actions[0] ).toEqual({
            type: types.uiSetError,
            payload: 'email is required'
        })

    });

    test('should show alert box with message', () => {

        // * note: actions dont modify state on testing cause reducer wont work

        const initState = {
            auth: {},
            ui: {
                loading: false,
                msgError: 'Email no es correcto' // err message
            }
        }

        const store = mockStore(initState);
        // we only simulate dispatch when we dont need it
        // store.dispatch = jest.fn();

        const wrapper = mount(
            <Provider store={ store }>
                <MemoryRouter>
                    <RegisterScreen />
                </MemoryRouter>
            </Provider>
        )

        expect( wrapper.find('.auth__alert-error').exists() ).toBe(true);
        expect( wrapper.find('.auth__alert-error').text().trim() ).toBe( initState.ui.msgError );

    });

});