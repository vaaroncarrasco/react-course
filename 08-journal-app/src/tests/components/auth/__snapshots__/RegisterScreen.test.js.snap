// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RegisterScreen /> should match snapshot 1`] = `
Array [
  <h3
    className="auth__title"
  >
    Register
  </h3>,
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="auth__input"
      name="name"
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value="aaron"
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="email"
      onChange={[Function]}
      placeholder="Email"
      type="text"
      value="aaron@aaron.com"
    />
    <input
      className="auth__input"
      name="password"
      onChange={[Function]}
      placeholder="Password"
      type="password"
      value="123456789"
    />
    <input
      className="auth__input"
      name="password2"
      onChange={[Function]}
      placeholder="Confirm password"
      type="password"
      value="123456789"
    />
    <button
      className="btn btn-primary btn-block mb-5"
      type="submit"
    >
      Register
    </button>
    <a
      className="link"
      href="/auth/login"
      onClick={[Function]}
    >
      Already registered?
    </a>
  </form>,
]
`;
